synapse/federation/transport/server/_base.py

```diff
@@ -309,59 +311,69 @@ async def new_func(
+                origin_context = context_from_request(request)
+
+            if origin_context:
+                local_servlet_span = get_active_span()
+                # Create a span which uses the `origin_context` as a parent
+                # so we can see how the incoming payload was processed while
+                # we're looking at the outgoing trace. Since the parent is set
+                # to a remote span (from the origin), it won't show up in the
+                # local trace which is why we create another span below for the
+                # local trace. A span can only have one parent so we have to
+                # create two separate ones.
+                remote_parent_span = start_span(
                     "incoming-federation-request",
-                    child_of=context,
-                    contexts=(servlet_span,),
-                    start_time=processing_start_time,
+                    context=origin_context,
+                    # Cross-link back to the local trace so we can jump
+                    # to the incoming side from the remote origin trace.
+                    links=[Link(local_servlet_span.get_span_context())]
+                    if local_servlet_span
+                    else None,
                 )
 
+                # Create a local span to appear in the local trace
+                local_parent_span_cm = start_active_span(
+                    "process-federation-request",
+                    # Cross-link back to the remote outgoing trace so we can
+                    # jump over there.
+                    links=[Link(remote_parent_span.get_span_context())],
+                )
             else:
-                # just use our context as a parent
-                scope = start_active_span(
-                    "incoming-federation-request",
+                # Otherwise just use our local active servlet context as a parent
+                local_parent_span_cm = start_active_span(
+                    "process-federation-request",
                 )
 
-            try:
-                with scope:
-                    if origin and self.RATELIMIT:
-                        with ratelimiter.ratelimit(origin) as d:
-                            await d
-                            if request._disconnected:
-                                logger.warning(
-                                    "client disconnected before we started processing "
-                                    "request"
-                                )
-                                return None
-                            response = await func(
-                                origin, content, request.args, *args, **kwargs
+                # Don't need to record anything for the remote because no remote
+                # trace context given.
+                remote_parent_span = create_non_recording_span()
+
+            remote_parent_span_cm = use_span(remote_parent_span, end_on_exit=True)
```

Comment:
Instead of needing to use `start_span`/`use_span`, is there a way to get the `__enter__` span value from the `ContextManager` that `start_active_span` returns?